File: sample_apps_reinvent2021/python/utils/WriteUtil.py
#!/usr/bin/python

import time
from utils.Constant import HT_TTL_HOURS, CT_TTL_DAYS


class WriteUtil:
    def __init__(self, client):
        self.client = client

    def create_database(self, database_name):
        print("Creating Database")
        try:
            self.client.create_database(DatabaseName=database_name)
            print("Database [%s] created successfully." % database_name)
        except self.client.exceptions.ConflictException:
            print("Database [%s] exists. Skipping database creation" % database_name)
        except Exception as err:
            print("Create database failed:", err)
            raise err

    def create_table(
        self,
        database_name,
        table_name,
        enable_magnetic_store_writes=True,
        magnetic_store_s3_bucket=None,
        memory_retention=HT_TTL_HOURS,
        magnetic_retention=CT_TTL_DAYS
    ):
        print("Creating table")
        retention_properties = {
            'MemoryStoreRetentionPeriodInHours': memory_retention,
            'MagneticStoreRetentionPeriodInDays': magnetic_retention
        }
        magnetic_store_write_properties = {
            'EnableMagneticStoreWrites': enable_magnetic_store_writes
        }
        if magnetic_store_s3_bucket:
            magnetic_store_write_properties['MagneticStoreRejectedDataLocation'] = {
                'S3Configuration': {
                    'BucketName': magnetic_store_s3_bucket,
                    'EncryptionOption': "SSE_S3"
                }
            }

        try:
            self.client.create_table(DatabaseName=database_name, TableName=table_name,
                                     RetentionProperties=retention_properties,
                                     MagneticStoreWriteProperties=magnetic_store_write_properties)
            print("Table [%s] successfully created." % table_name)
        except self.client.exceptions.ConflictException:
            print("Table [%s] exists on database [%s]. Skipping table creation" % (
                table_name, database_name))
        except Exception as err:
            print("Create table failed:", err)
            raise err

    def update_table(
        self,
        database_name,
        table_name,
        memory_retention=HT_TTL_HOURS,
        magnetic_retention=CT_TTL_DAYS,
    ):
        retention_properties = {
            'MemoryStoreRetentionPeriodInHours': memory_retention,
            'MagneticStoreRetentionPeriodInDays': magnetic_retention
        }

        try:
            self.client.update_table(DatabaseName=database_name, TableName=table_name,
                                     RetentionProperties=retention_properties)
            print("Table updated.")
        except Exception as err:
            print("Update table failed:", err)

    def delete_table(self, database_name, table_name):
        print(f"Deleting Table: {table_name}")
        try:
            result = self.client.delete_table(DatabaseName=database_name, TableName=table_name)
            if result and result['ResponseMetadata']:
                print("Delete table status [%s]" % result['ResponseMetadata']['HTTPStatusCode'])
        except self.client.exceptions.ResourceNotFoundException:
            print("Table [%s] doesn't exist" % table_name)
        except Exception as err:
            # Not raising an exception here as we want other cleanup to continue
            print("Delete table failed:", err)

    def delete_database(self, database_name):
        print(f"Deleting Database: {database_name}")
        try:
            result = self.client.delete_database(DatabaseName=database_name)
            if result and result['ResponseMetadata']:
                print("Delete database status [%s]" % result['ResponseMetadata']['HTTPStatusCode'])
        except self.client.exceptions.ResourceNotFoundException:
            print("database [%s] doesn't exist" % database_name)
        except Exception as err:
            # Not raising an exception here as we want other cleanup to continue
            print("Delete database failed:", err)

    @staticmethod
    def print_rejected_records_exceptions(err):
        print("RejectedRecords: ", err)
        for rr in err.response["RejectedRecords"]:
            print("Rejected Index " + str(rr["RecordIndex"]) + ": " + rr["Reason"])
            if "ExistingVersion" in rr:
                print("Rejected record existing version: ", rr["ExistingVersion"])

    @staticmethod
    def current_milli_time():
        return str(int(round(time.time() * 1000)))

    def describe_database(self, database_name):
        print("Describing database")
        try:
            result = self.client.describe_database(DatabaseName=database_name)
            print("Database [%s] has id [%s]" % (database_name, result['Database']['Arn']))
        except self.client.exceptions.ResourceNotFoundException:
            print("Database doesn't exist")
        except Exception as err:
            print("Describe database failed:", err)

    def update_database(self, database_name, kms_id):
        print("Updating database")
        try:
            result = self.client.update_database(DatabaseName=database_name, KmsKeyId=kms_id)
            print("Database [%s] was updated to use kms [%s] successfully" % (database_name,
                                                                              result['Database']['KmsKeyId']))
        except self.client.exceptions.ResourceNotFoundException:
            print("Database doesn't exist")
        except Exception as err:
            print("Update database failed:", err)

    def list_databases(self):
        print("Listing databases")
        try:
            result = self.client.list_databases(MaxResults=5)
            self.print_databases(result['Databases'])
            next_token = result.get('NextToken', None)
            while next_token:
                result = self.client.list_databases(NextToken=next_token, MaxResults=5)
                self.print_databases(result['Databases'])
                next_token = result.get('NextToken', None)
        except Exception as err:
            print("List databases failed:", err)

    def describe_table(self, database_name, table_name):
        print("Describing table")
        try:
            result = self.client.describe_table(DatabaseName=database_name, TableName=table_name)
            print("Table [%s] has id [%s]" % (table_name, result['Table']['Arn']))
        except self.client.exceptions.ResourceNotFoundException:
            print("Table doesn't exist")
        except Exception as err:
            print("Describe table failed:", err)

    def list_tables(self, database_name):
        print("Listing tables")
        try:
            result = self.client.list_tables(DatabaseName=database_name, MaxResults=5)
            self.print_tables(result['Tables'])
            next_token = result.get('NextToken', None)
            while next_token:
                result = self.client.list_tables(DatabaseName=database_name,
                                                 NextToken=next_token, MaxResults=5)
                self.print_tables(result['Tables'])
                next_token = result.get('NextToken', None)
        except Exception as err:
            print("List tables failed:", err)

    @staticmethod
    def print_tables(tables):
        for table in tables:
            print(table['TableName'])

    @staticmethod
    def print_databases(databases):
        for database in databases:
            print(database['DatabaseName'])

File: sample_apps/java/src/main/java/com/amazonaws/services/timestream/utils/WriteUtil.java
package com.amazonaws.services.timestream.utils;

import java.util.List;

import com.amazonaws.services.timestreamwrite.AmazonTimestreamWrite;
import com.amazonaws.services.timestreamwrite.model.ConflictException;
import com.amazonaws.services.timestreamwrite.model.CreateDatabaseRequest;
import com.amazonaws.services.timestreamwrite.model.CreateTableRequest;
import com.amazonaws.services.timestreamwrite.model.Database;
import com.amazonaws.services.timestreamwrite.model.DeleteDatabaseRequest;
import com.amazonaws.services.timestreamwrite.model.DeleteDatabaseResult;
import com.amazonaws.services.timestreamwrite.model.DeleteTableRequest;
import com.amazonaws.services.timestreamwrite.model.DeleteTableResult;
import com.amazonaws.services.timestreamwrite.model.DescribeDatabaseRequest;
import com.amazonaws.services.timestreamwrite.model.DescribeDatabaseResult;
import com.amazonaws.services.timestreamwrite.model.DescribeTableRequest;
import com.amazonaws.services.timestreamwrite.model.DescribeTableResult;
import com.amazonaws.services.timestreamwrite.model.ListDatabasesRequest;
import com.amazonaws.services.timestreamwrite.model.ListDatabasesResult;
import com.amazonaws.services.timestreamwrite.model.ListTablesRequest;
import com.amazonaws.services.timestreamwrite.model.ListTablesResult;
import com.amazonaws.services.timestreamwrite.model.Record;
import com.amazonaws.services.timestreamwrite.model.ResourceNotFoundException;
import com.amazonaws.services.timestreamwrite.model.RetentionProperties;
import com.amazonaws.services.timestreamwrite.model.Table;
import com.amazonaws.services.timestreamwrite.model.UpdateDatabaseRequest;
import com.amazonaws.services.timestreamwrite.model.UpdateDatabaseResult;
import com.amazonaws.services.timestreamwrite.model.UpdateTableRequest;
import com.amazonaws.services.timestreamwrite.model.ValidationException;
import com.amazonaws.services.timestreamwrite.model.WriteRecordsRequest;
import com.amazonaws.services.timestreamwrite.model.WriteRecordsResult;

import static com.amazonaws.services.timestream.utils.Constants.CT_TTL_DAYS;
import static com.amazonaws.services.timestream.utils.Constants.HT_TTL_HOURS;

public class WriteUtil {

    AmazonTimestreamWrite amazonTimestreamWrite;

    public WriteUtil(AmazonTimestreamWrite client) {
        this.amazonTimestreamWrite = client;
    }

    public void createDatabase(String databaseName) {
        System.out.println("Creating database");
        CreateDatabaseRequest request = new CreateDatabaseRequest();
        request.setDatabaseName(databaseName);
        try {
            amazonTimestreamWrite.createDatabase(request);
            System.out.println("Database [" + databaseName + "] created successfully");
        } catch (ConflictException e) {
            System.out.println("Database [" + databaseName + "] exists. Skipping database creation");
        }
    }

    public void describeDatabase(String databaseName) {
        System.out.println("Describing database");
        final DescribeDatabaseRequest describeDatabaseRequest = new DescribeDatabaseRequest();
        describeDatabaseRequest.setDatabaseName(databaseName);
        try {
            DescribeDatabaseResult result = amazonTimestreamWrite.describeDatabase(describeDatabaseRequest);
            final Database databaseRecord = result.getDatabase();
            final String databaseId = databaseRecord.getArn();
            System.out.println("Database " + databaseName + " has id " + databaseId);
        } catch (final Exception e) {
            System.out.println("Database doesn't exist = " + e);
            throw e;
        }
    }

    public void listDatabases() {
        System.out.println("Listing databases");
        ListDatabasesRequest request = new ListDatabasesRequest();
        ListDatabasesResult result = amazonTimestreamWrite.listDatabases(request);
        final List<Database> databases = result.getDatabases();
        printDatabases(databases);

        String nextToken = result.getNextToken();
        while (nextToken != null && !nextToken.isEmpty()) {
            request.setNextToken(nextToken);
            ListDatabasesResult nextResult = amazonTimestreamWrite.listDatabases(request);
            final List<Database> nextDatabases = nextResult.getDatabases();
            printDatabases(nextDatabases);
            nextToken = nextResult.getNextToken();
        }
    }

    public void createTable(String databaseName, String tableName) {
        System.out.println("Creating table");
        CreateTableRequest createTableRequest = new CreateTableRequest();
        createTableRequest.setDatabaseName(databaseName);
        createTableRequest.setTableName(tableName);
        final RetentionProperties retentionProperties = new RetentionProperties()
                .withMemoryStoreRetentionPeriodInHours(HT_TTL_HOURS)
                .withMagneticStoreRetentionPeriodInDays(CT_TTL_DAYS);
        createTableRequest.setRetentionProperties(retentionProperties);

        try {
            amazonTimestreamWrite.createTable(createTableRequest);
            System.out.println("Table [" + tableName + "] successfully created.");
        } catch (ConflictException e) {
            System.out.println("Table [" + tableName + "] exists on database [" + databaseName + "] . Skipping database creation");
        }
    }

    public void updateTable(String databaseName, String tableName) {
        System.out.println("Updating table");
        UpdateTableRequest updateTableRequest = new UpdateTableRequest();
        updateTableRequest.setDatabaseName(databaseName);
        updateTableRequest.setTableName(tableName);

        final RetentionProperties retentionProperties = new RetentionProperties()
                .withMemoryStoreRetentionPeriodInHours(HT_TTL_HOURS)
                .withMagneticStoreRetentionPeriodInDays(CT_TTL_DAYS);

        updateTableRequest.setRetentionProperties(retentionProperties);

        amazonTimestreamWrite.updateTable(updateTableRequest);
        System.out.println("Table updated");
    }

    public void describeTable(String databaseName, String tableName) {
        System.out.println("Describing table");
        final DescribeTableRequest describeTableRequest = new DescribeTableRequest();
        describeTableRequest.setDatabaseName(databaseName);
        describeTableRequest.setTableName(tableName);
        try {
            DescribeTableResult result = amazonTimestreamWrite.describeTable(describeTableRequest);
            String tableId = result.getTable().getArn();
            System.out.println("Table " + tableName + " has id " + tableId);
        } catch (final Exception e) {
            System.out.println("Table " + tableName + " doesn't exist = " + e);
            throw e;
        }
    }

    public void listTables(String databaseName) {
        System.out.println("Listing tables");
        ListTablesRequest request = new ListTablesRequest();
        request.setDatabaseName(databaseName);
        ListTablesResult result = amazonTimestreamWrite.listTables(request);
        printTables(result.getTables());

        String nextToken = result.getNextToken();
        while (nextToken != null && !nextToken.isEmpty()) {
            request.setNextToken(nextToken);
            ListTablesResult nextResult = amazonTimestreamWrite.listTables(request);

            printTables(nextResult.getTables());
            nextToken = nextResult.getNextToken();
        }
    }

    public void deleteTable(String databaseName, String tableName) {
        System.out.println("Deleting table");
        final DeleteTableRequest deleteTableRequest = new DeleteTableRequest();
        deleteTableRequest.setDatabaseName(databaseName);
        deleteTableRequest.setTableName(tableName);
        try {
            DeleteTableResult result =
                    amazonTimestreamWrite.deleteTable(deleteTableRequest);
            System.out.println("Delete table status: " + result.getSdkHttpMetadata().getHttpStatusCode());
        } catch (final ResourceNotFoundException e) {
            System.out.println("Table " + tableName + " doesn't exist = " + e);
            throw e;
        } catch (final Exception e) {
            System.out.println("Could not delete table " + tableName + " = " + e);
            throw e;
        }
    }

    public void deleteDatabase(String databaseName) {
        System.out.println("Deleting database");
        final DeleteDatabaseRequest deleteDatabaseRequest = new DeleteDatabaseRequest();
        deleteDatabaseRequest.setDatabaseName(databaseName);
        try {
            DeleteDatabaseResult result =
                    amazonTimestreamWrite.deleteDatabase(deleteDatabaseRequest);
            System.out.println("Delete database status: " + result.getSdkHttpMetadata().getHttpStatusCode());
        } catch (final ResourceNotFoundException e) {
            System.out.println("Database " + databaseName + " doesn't exist = " + e);
            throw e;
        } catch (final Exception e) {
            System.out.println("Could not delete Database " + databaseName + " = " + e);
            throw e;
        }
    }

    public void updateDatabase(String databaseName, String kmsId) {
        System.out.println("Updating kmsId to " + kmsId);
        UpdateDatabaseRequest request = new UpdateDatabaseRequest();
        request.setDatabaseName(databaseName);
        request.setKmsKeyId(kmsId);
        try {
            UpdateDatabaseResult result = amazonTimestreamWrite.updateDatabase(request);
            System.out.println("Update Database complete");
        } catch (final ValidationException e) {
            System.out.println("Update database failed:");
            e.printStackTrace();
        } catch (final ResourceNotFoundException e) {
            System.out.println("Database " + databaseName + " doesn't exist = " + e);
        } catch (final Exception e) {
            System.out.println("Could not update Database " + databaseName + " = " + e);
            throw e;
        }
    }

    public void submitBatch(String databaseName, String tableName, List<Record> records, int counter) {
        WriteRecordsRequest writeRecordsRequest = new WriteRecordsRequest()
                .withDatabaseName(databaseName)
                .withTableName(tableName);
        writeRecordsRequest.setRecords(records);

        try {
            WriteRecordsResult writeRecordsResult = amazonTimestreamWrite.writeRecords(writeRecordsRequest);
            System.out.println("Processed " + counter + " records. WriteRecords Status: " +
                    writeRecordsResult.getSdkHttpMetadata().getHttpStatusCode());
        } catch (Exception e) {
            System.out.println("Error: " + e);
        }
    }

    private void printDatabases(List<Database> databases) {
        for (Database db : databases) {
            System.out.println(db.getDatabaseName());
        }
    }

    private void printTables(List<Table> tables) {
        for (Table table : tables) {
            System.out.println(table.getTableName());
        }
    }
}
